auto_update.go

```diff
@@ -18,62 +15,47 @@ import (
 )
 
 const locallyBuilt = "(Locally Built)"
+const autoUpdateFile = "TGFAutoUpdate"
+
+// RunnerUpdater allows flexibility for testing
+type RunnerUpdater interface {
+	Debug(format string, args ...interface{})
+	GetUpdateVersion() (string, error)
+	GetLastRefresh(file string) time.Duration
+	SetLastRefresh(file string)
+	ShouldUpdate() bool
+	Run() int
+	Restart() int
+}
 
 // RunWithUpdateCheck checks if an update is due, checks if current version is outdated and performs update if needed
-func (c *TGFConfig) RunWithUpdateCheck(lastRefresh func(image string) time.Duration) int {
-	app := c.tgf
-	const autoUpdateFile = "TGFAutoUpdate"
-
-	if app.AutoUpdateSet {
-		if app.AutoUpdate {
-			app.Debug("Auto update is forced. Checking version...")
-		} else {
-			app.Debug("Auto update is force disabled. Bypassing update version check.")
-			return c.Run()
-		}
-	} else {
-		if !c.AutoUpdate {
-			app.Debug("Auto update is disabled in the config. Bypassing update version check.")
-			return c.Run()
-		}
-		if lastRefresh(autoUpdateFile) < c.AutoUpdateDelay {
-			app.Debug("Less than %v since last check. Bypassing update version check.", c.AutoUpdateDelay.String())
-			return c.Run()
-		}
-	}
-
-	app.Debug("Comparing local and latest versions...")
-	touchImageRefresh(autoUpdateFile)
+func RunWithUpdateCheck(c RunnerUpdater) int {
 
-	latestVersionString := c.UpdateVersion
-	if latestVersionString == "" {
-		fetchedVersion, err := getLatestVersion()
-		if err != nil {
-			printError("Error getting latest version: %v", err)
-			return c.Run()
-		}
-		latestVersionString = fetchedVersion
+	if !c.ShouldUpdate() {
+		return c.Run()
 	}
 
-	latestVersion, err := semver.Make(latestVersionString)
+	c.Debug("Comparing local and latest versions...")
+	c.SetLastRefresh(autoUpdateFile)
```

Comment:
Why does this need to be exposed? It could run at the end of the `ShouldUpdate`

Reply:
Because it's a side effect, I think I should mock it to avoid actually creating a file.